# Iris DataSet
En este problema se dará solución al problema de clasificación multiclase al dataset Iris mediante la apliación de 3 técnicas de clasificación:

- Logistic Regression One vs. All, 
- Logistic Regression One vs. One y 
- Softmax Regression

Primeramente, se realiza la importación de todas las librerías necesarias, incluyendo la clase datasets de la ñibrería *sklearn*, de donde se tomará el dataset

In [28]:
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn import datasets
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

Seguidamente, examinaremos un poco el dataset con el que trabajaremos, detallando los parámetros que tiene, las etiquetas que maneja con el fin de conocer un poco sobre su constitución.

In [29]:
#Cargamos el dataset
iris = datasets.load_iris() 
#revisamos su forma
qty = np.shape(iris.target)
features = iris.feature_names
targets = iris.target_names

#Checamos la cantidad de datos que hay por clase
unique_targets, counts = np.unique(iris.target, return_counts=True)

print(f'Los parámetros son: {features}\n')
print(f'Las clases son: {targets}\n')
print(f'La cantidad de datos son: {qty[0]}\n')
print('La distribución que sigue es:')
[print(f'Clase {unique_targets[i]} : {counts[i]}') for i in range(np.shape(unique_targets)[0])]


Los parámetros son: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

Las clases son: ['setosa' 'versicolor' 'virginica']

La cantidad de datos son: 150

La distribución que sigue es:
Clase 0 : 50
Clase 1 : 50
Clase 2 : 50


[None, None, None]

Podemos resumir lo que acabamos de hacer en las siguientes ideas:

- El dataset cuenta con 150 datos.
- Hay 3 clases diferentes: setosa, versicolor y virginica.
- Cada clase cuenta con 50 datos.
- Los parámetros que permiten clasificar un sujeto en una clase son: largo del sepal, ancho del sepal, largo del pétalo y ancho del pétalo; todos medidos en cm.

Lo que resta ahora es dividir el dataset en datos de entrenamiento y evaluación. Es importane mencionar las sigueintes dos consideraciones:
- Debido a reducido tamaño del dataset, los datos de validación y de evaluación serán los mismos **(REVISAR)**
- Para tener una buena precisión, la división de los datos debe de ser equitativa.

In [38]:
ts = 0.2 #Test size
seed = 34 #semilla

train_iris, test_iris, train_target, test_target = train_test_split(iris.data, iris.target, test_size=ts, 
                                                                    random_state=seed, shuffle=True,
                                                                    stratify=iris.target)

#Verificando la división equitativa
#unique, count = np.unique(train_target, return_counts=True)
#print(f'{unique}, {count}')